### PR TITLE
[SPARK-31176][SQL] Remove support for 'e'/'c' as datetime pattern character

### PR DESCRIPTION
**Pull request by yaooqinn**
### What changes were proposed in this pull request?

The meaning of 'u' was day number of the week in SimpleDateFormat, it was changed to year in DateTimeFormatter. Now we keep the old meaning of 'u' by substituting 'u' to 'e' internally and use DateTimeFormatter to parse the pattern string. In DateTimeFormatter, the 'e' and 'c' also represents day-of-week. e.g.

```sql
select date_format(timestamp '2019-10-06', 'yyyy-MM-dd uuuu');
select date_format(timestamp '2019-10-06', 'yyyy-MM-dd uuee');
select date_format(timestamp '2019-10-06', 'yyyy-MM-dd eeee');
```
Because of the substitution, they all goes to `.... eeee` silently. The users may congitive problems of their meanings, so we should mark them as illegal pattern characters to stay the same as before.

This pr move the method `convertIncompatiblePattern` from `DatetimeUtils` to `DateTimeFormatterHelper` object, since it is quite specific for `DateTimeFormatterHelper` class.
And 'e' and 'c' char checking in this method.

Besides,`convertIncompatiblePattern` has a bug that will lose the last `'` if it ends with it, this pr fixes this too. e.g.

```sql
spark-sql> select date_format(timestamp "2019-10-06", "yyyy-MM-dd'S'");
20/03/18 11:19:45 ERROR SparkSQLDriver: Failed in [select date_format(timestamp "2019-10-06", "yyyy-MM-dd'S'")]
java.lang.IllegalArgumentException: Pattern ends with an incomplete string literal: uuuu-MM-dd'S

spark-sql> select to_timestamp("2019-10-06S", "yyyy-MM-dd'S'");
NULL
```
### Why are the changes needed?

avoid vagueness
bug fix

### Does this PR introduce any user-facing change?

no, these are not  exposed yet

### How was this patch tested?

add ut

### CODE REVIEW

**Comment by yaooqinn** [comment]
cc @cloud-fan @maropu @MaxGekk @xuanyuanking @dongjoon-hyun thanks.

File: sql/core/src/test/resources/sql-tests/inputs/datetime.sql [hunk 1]
```diff
@@ -100,3 +100,10 @@ select to_timestamp("12.12342019-10-06S10:11", "ss.SSSSyyyy-MM-dd'S'HH:mm");
 select to_timestamp("12.1232019-10-06S10:11", "ss.SSSSyyyy-MM-dd'S'HH:mm");
 select to_timestamp("12.1232019-10-06S10:11", "ss.SSSSyy-MM-dd'S'HH:mm");
 select to_timestamp("12.1234019-10-06S10:11", "ss.SSSSy-MM-dd'S'HH:mm");
+
+select to_timestamp("2019-10-06S", "yyyy-MM-dd'S'");
+select to_timestamp("S2019-10-06", "'S'yyyy-MM-dd");
+
+select date_format(timestamp '2019-10-06', 'yyyy-MM-dd eeee');
+select date_format(timestamp '2019-10-06', 'yyyy-MM-dd cccc');
+select date_format(timestamp '2019-10-06', 'yyyy-MM-dd uuuu');
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/inputs/datetime.sql]: we can add a test of `uuee` to demonstrate the problem?

Comment (cloud-fan) [reply sql/core/src/test/resources/sql-tests/inputs/datetime.sql]: We should also highlight this problem in PR description.

Comment (yaooqinn) [reply sql/core/src/test/resources/sql-tests/inputs/datetime.sql]: updated. thanks.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeFormatterHelper.scala [hunk 2]
```diff
@@ -116,7 +116,8 @@ private object DateTimeFormatterHelper {
   def createBuilderWithVarLengthSecondFraction(
       pattern: String): DateTimeFormatterBuilder = {
     val builder = createBuilder()
-    pattern.split("'").zipWithIndex.foreach {
+    val patternSplits = pattern.split("'")
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeFormatterHelper.scala]: unnecessary change?

**Review by cloud-fan — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeFormatterHelper.scala (hunk 2 above)
Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeFormatterHelper.scala]: thanks.

**Comment by yaooqinn** [comment]
retest this please

**Comment by cloud-fan** [comment]
thanks, merging to master/3.0!

**Comment by xuanyuanking** [comment]
Late LGTM, thanks for the fix!
